### DISCUSSION TITLE
"time" keyword output not redirected

### DISCUSSION THREAD

**Discussion by nap2001** [fish-shell/fish-shell#10755] [Q&A] [answered]
I frequently use "&| tee" to capture output from long running commands to a log file. I just discovered that the output of fish shell's time keyword can not be captured. I am wondering why? Are there any other instances where command output to my terminal session can not be captured or redirected? Just curious.

**Comment by faho** [comment, marked as answer]
> I just discovered that the output of fish shell's time keyword can not be captured. I am wondering why? 

Because how would you measure the time for commands with redirections otherwise?

The trade-off is either `time foo >/dev/null` sends the time output to /dev/null as well, or time isn't redirected. Since timings may change with redirections (isatty, terminals holding up output, ...), this is a reasonably important thing to measure, and so we chose the latter. It shares this with e.g. bash's and zsh's time.

We *could* add some way to redirect time specifically, but tbh it's mostly for seeing interactively how long something takes, not a rigorous benchmarking tool.

(and redirecting *elsewhere* would interleave time's output with that of the command, which is also typically not what you want)

> Are there any other instances where command output to my terminal session can not be captured or redirected

None spring to mind, no.

**Reply by nap2001** [reply]
Thank you for the answer. While I'm not using it for benchmarking, I really prefer the output provided  by fish shell's time and would be okay with the overhead of redirection being included. I'm looking to have measurements for the complete command line I'm typing so redirection would be part of that. 

I just hoped that something like parenthesis or begin end would allow one to capture its output.  I can use /usr/bin/time with "&| tee" so I will probably settle for using its inferior reporting.

**Reply by sbirch-fl** [reply]
I suppose you could run `time` in a subshell. E.g.,
```
> fish -c 'time md5sum /bin/gz*' &| nl
     1  9ede9e20fdee2e757019cba41d9e9d10  /bin/gzexe
     2  c1d67f6d74bf7c0e16f1f218ead05e2f  /bin/gzip

     3  ________________________________________________________
     4  Executed in    2.37 millis    fish           external
     5     usr time    0.30 millis  299.00 micros    0.00 millis
     6     sys time    1.82 millis  449.00 micros    1.38 millis
```
cf.,
```
> time md5sum /bin/gz* &| nl
     1  9ede9e20fdee2e757019cba41d9e9d10  /bin/gzexe
     2  c1d67f6d74bf7c0e16f1f218ead05e2f  /bin/gzip

________________________________________________________
Executed in    6.47 millis    fish           external
   usr time    1.14 millis    0.22 millis    0.92 millis
   sys time    9.12 millis    3.04 millis    6.08 millis
```

**Comment by krobelus** [comment]
> I suppose you could run `time` in a subshell.

Yep.

A POSIX alternative is to add redirections to special builtins like "exec 2>fifo".
That will redirect the shell's error stream itself.
But it obviously doesn't work with anonymous pipes.
